### PR TITLE
Never create allocas for indirect function arguments

### PR DESCRIPTION
**Pull request by dotdash**
The assertion that argument debuginfo always has to be assigned to
allocas only exists in rustc, not in LLVM. The reason for that assertion
might have been that we used to created bad debuginfo for closures which
did indeed lead to errors when we skipped the alloca, but this was fixed
in commit 218eccf "Fix de-deduplication for closure debuginfo".

So now we can always skip the alloca for indirect arguments, even when
generating debuginfo.

### CODE REVIEW

File: src/librustc_trans/trans/base.rs [hunk 1]
```diff
@@ -1370,9 +1370,9 @@ pub fn create_datums_for_fn_args<'a, 'tcx>(mut bcx: Block<'a, 'tcx>,
...
     let mut idx = fcx.arg_offset() as c_uint;
     for (i, &arg_ty) in arg_tys.iter().enumerate() {
+        let mut indirect_arg = type_of::arg_is_indirect(bcx.ccx(), arg_ty);
         let arg_datum = if !has_tupled_arg || i < arg_tys.len() - 1 {
-            if type_of::arg_is_indirect(bcx.ccx(), arg_ty)
-                    && bcx.sess().opts.debuginfo != FullDebugInfo {
+            if indirect_arg {
                 // Don't copy an indirect argument to an alloca, the caller
                 // already put it in a temporary alloca and gave it up, unless
                 // we emit extra-debug-info, which requires local allocas :(.
```
Comment (rkruppe) [inline src/librustc_trans/trans/base.rs]: The last part of this comment is outdated now, isn't it?

Comment (dotdash) [inline src/librustc_trans/trans/base.rs]: Right!

**Comment by michaelwoerister** [comment]
Interesting. I remember running into lots of problems when not having allocas and LLVM didn't really have any documentation on it's requirement in this respect. But that was quite some time and LLVM updates ago and before unboxed closures were a thing in Rust. Maybe things have changed.
I assume this PR passes make check?

**Comment by dotdash** [comment]
Yes, it did pass make check for me. Had to wait for the new snapshot that
has the mentioned fix because without that it would fail in check-stage1
Am 13.08.2015 23:23 schrieb "Michael Woerister" <notifications@github.com>:

**Comment by michaelwoerister** [comment]
OK, let's give it a try `:)`

**Comment by dotdash** [comment]
@michaelwoerister Any idea what might be wrong here? gdb has no problem with those changes, and AFAICT, this PR makes rustc behave the same as clang, so I don't really know where to look.

**Comment by michaelwoerister** [comment]
I'll see if I can find out something over the weekend.

**Comment by michaelwoerister** [comment]
OK, so I can reproduce this error locally for 32bit builds and it seems that the DWARF information does not contain the entries for some parameters. E.g. in associated-types.rs for `assoc_tuple<>()` there is no entry for `arg`:
```
0x00000206:         TAG_subprogram [12] *
                     AT_low_pc( 0x00001c00 )
                     AT_high_pc( 0x00001c14 )
                     AT_frame_base( ebp )
                     AT_MIPS_linkage_name( "_ZN16associated_types16assoc_tuple<i32>E" )
                     AT_name( "assoc_tuple<i32>" )
                     AT_decl_file( "/Users/mw/rust-mac/src/test/debuginfo/associated-types.rs" )
                     AT_decl_line( 125 )

0x0000021b:             TAG_template_type_parameter [14]  
                         AT_type( {0x000002a7} ( i32 ) )
                         AT_name( "T" )

0x00000224:             NULL
```
In other cases, it is there as expected:
```
0x000001a8:         TAG_subprogram [12] *
                     AT_low_pc( 0x00001bb0 )
                     AT_high_pc( 0x00001bd0 )
                     AT_frame_base( ebp )
                     AT_MIPS_linkage_name( "_ZN16associated_types14assoc_arg<i32>E" )
                     AT_name( "assoc_arg<i32>" )
                     AT_decl_file( "/Users/mw/rust-mac/src/test/debuginfo/associated-types.rs" )
                     AT_decl_line( 117 )

0x000001bd:             TAG_formal_parameter [10]  
                         AT_location( fbreg -16 )
                         AT_name( "arg" )
                         AT_decl_file( "/Users/mw/rust-mac/src/test/debuginfo/associated-types.rs" )
                         AT_decl_line( 117 )
                         AT_type( {0x000002ae} ( i64 ) )

0x000001cb:             TAG_template_type_parameter [14]  
                         AT_type( {0x000002a7} ( i32 ) )
                         AT_name( "T" )

0x000001d4:             NULL
```
In the 64bit version, the entry does show up:
```
0x0000024a:         TAG_subprogram [12] *
                     AT_low_pc( 0x0000000100000b40 )
                     AT_high_pc( 0x0000000100000b5b )
                     AT_frame_base( rbp )
                     AT_MIPS_linkage_name( "_ZN16associated_types16assoc_tuple<i32>E" )
                     AT_name( "assoc_tuple<i32>" )
                     AT_decl_file( "/Users/mw/rust-mac/src/test/debuginfo/associated-types.rs" )
                     AT_decl_line( 125 )

0x00000267:             TAG_formal_parameter [13]  
                         AT_location( 0x00000039
                            0x0000000100000b40 - 0x0000000100000b50: rdi+0
                            0x0000000100000b50 - 0x0000000100000b5b: rbp-16, deref )
                         AT_name( "arg" )
                         AT_decl_file( "/Users/mw/rust-mac/src/test/debuginfo/associated-types.rs" )
                         AT_decl_line( 125 )
                         AT_type( {0x0000031d} ( (i32, i64) ) )

0x00000276:             TAG_template_type_parameter [14]  
                         AT_type( {0x0000030f} ( i32 ) )
                         AT_name( "T" )

0x0000027f:             NULL
```
Is it possible that the 32bit code generator will just remove the variable altogether while the 64bit one doesn't?

**Comment by dotdash** [comment]
Weird, I can reproduce this on Linux i686 now, too. Not sure why it didn't happen before. And, I can also reproduce the problem with clang.

**Comment by dotdash** [comment]
@michaelwoerister I figured this out now. It's a limitation of FastISel. On X86_64, the pointer is in a register, but on x86 it's on the stack. And apparently handling the value on the stack is something FastISel can't do yet.

https://github.com/llvm-mirror/llvm/blob/master/lib/CodeGen/SelectionDAG/FastISel.cpp#L1173

That is triggered for x86 but not for x86_64.

I'm not sure how we want to proceed here. Drop the patch or only skip the alloca when optimizations are turned on?

**Comment by michaelwoerister** [comment]
@dotdash Cool that you found out what the problem is in LLVM!
So your thinking is that if somebody switches on debuginfo+optimizations they can't rely on variable info being there anyway and thus we can risk losing the information during codegen, right? That would make sense to me. On the other hand, what is there a big detriment in always generating the alloca when debuginfo is turned on?
As this slightly complicates the code and leads to some unexpected effects when lowering from LLVM IR to machine code, on just some platforms, I'd be for dropping the patch -- unless you can name a good reason not to, of course `:)`

**Comment by dotdash** [comment]
@michaelwoerister Well, the information is only "guaranteed" to be lost when optimizations are turned off, and thus FastISel is turned on. With optimizations turned on, the debug info is possibly preserved.

Anyway, I agree on this slightly complicating the code. So until I find a less complicated way and the problem in LLVM gets fixed, I'll close this.

**Pull request closed without merging**